python/cudf/cudf/_libxx/join.pyx

```diff
@@ -0,0 +1,139 @@
+from cudf._libxx.lib cimport *
+from cudf._libxx.table cimport Table
+from cudf._libxx.includes.join cimport *
+from libcpp.vector cimport vector
+from libcpp.pair cimport pair
+
+from collections import OrderedDict
+
+cpdef join(lhs, rhs, left_on, right_on, how, method, left_index=False, right_index=False):
+    """
+    Call libcudf++ join for full outer, inner and left joins.
+    Returns a list of tuples [(column, valid, name), ...]
+    """
+
+    cdef vector[int] left_on_ind
+    cdef vector[int] right_on_ind
+    cdef vector[pair[int, int]] columns_in_common
+
+    # convert both left and right to c table objects
+    cdef Table c_lhs = lhs
+    cdef Table c_rhs = rhs
+
+    # create the two views that will be set later to include or not include the index column 
+    cdef table_view lhs_view
+    cdef table_view rhs_view
+
+    # the result columns are all the left columns (including common ones) 
+    # + all the right columns (excluding the common ones)
+    result_col_names = []
+
+    for name in lhs._data.keys():
+        result_col_names.append(name)
+    for name in rhs._data.keys():
+        if name not in lhs._data.keys():
+            result_col_names.append(name)
+
+    # keep track of where the desired index column will end up
+    result_index_pos = None
+    # if left_index or right_index is true, then both indices must be processed as join columns
+    if left_index or right_index:
+        # If one index is specified, it will be joined against some column from the other
+        # the result will be returned as a gather from the other column, and the index itself drops
+        # In addition, the opposite index ends up as the final index 
+        # Thus if either are true, we need to process both indices as join columns
+        lhs_view = c_lhs.view()
+        rhs_view = c_rhs.view()
+
+        left_join_cols = [lhs.index.name] + list(lhs._data.keys())
+        right_join_cols = [rhs.index.name] + list(rhs._data.keys())
```

Comment:
What if `lhs.index` or `rhs.index` is a MultiIndex here?

Reply:
As far as I could tell multiindex joins are not supported by the [legacy code](https://github.com/rapidsai/cudf/blob/branch-0.13/python/cudf/cudf/core/dataframe.py#L2451), so I assumed single indices for everything.

Reply:
Could we raise a `NotImplementedError` or something similar if index is a MultiIndex for the time being then?

Reply:
Added this.